### PR TITLE
add a system variable tiflash_replica_read

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [ ] v7.2 (TiDB 7.2 versions)
- [ ] v7.1 (TiDB 7.1 versions)
- [ ] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: system-variables.md [hunk 1]
```diff
@@ -4351,3 +4351,18 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
 - 类型：布尔型
 - 默认值：`ON`
 - 这个变量用于控制计算窗口函数时是否采用高精度模式。
+
+### tidb_tiflash_node_selection_policy
```
Comment (qiancai) [inline system-variables.md:4355]: system-variables.md 里的变量是按照字母顺序排列的，请调整下它的位置

File: system-variables.md [hunk 2]
```diff
@@ -4172,6 +4172,21 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
 - 理论上窗口函数的性能会随着该值的增加线性提升。但是如果设置的值超过实际的物理线程数，反而会导致性能下降。
 
+### tiflash_replica_read
+
+- 作用范围：SESSION | GLOBAL
+- 持久化至集群：是
+- 类型：枚举
+- 默认值："all_replicas"
+- 可选值："all_replicas"、"closest_adaptive"、"closest_replicas"
+- 该变量用于设置当查询需要使用 TiFlash 引擎时，TiFlash 节点的选择策略。
```
Comment (joey-yez) [inline system-variables.md:4182]:
```suggestion
- 该变量用于设置当查询需要使用 TiFlash 引擎时，TiFlash 副本的选择策略。
```

File: system-variables.md [hunk 3]
```diff
@@ -4172,6 +4172,21 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+  - "all_replicas" 表示使用所有可用节点进行分析计算。
```
Comment (joey-yez) [inline system-variables.md:4183]:
```suggestion
  - "all_replicas" 表示使用所有的 TiFlash 副本进行分析计算。
```

File: system-variables.md [hunk 4]
```diff
@@ -4172,6 +4172,21 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+  - "closest_adaptive" 表示使用与入口 TiDB 相同区域的节点。如果无法访问所有的 TiFlash 数据，则查询将通过与入口 TiDB 相同区域的节点访问来自其他区域的 TiFlash 节点。
+  - "closest_replicas" 表示仅使用与入口 TiDB 相同区域的节点。如果无法访问所有的 TiFlash 数据，则查询将报错。
```
Comment (joey-yez) [inline system-variables.md:4185]:
```suggestion
  - "closest_replicas" 表示仅使用与入口 TiDB 相同区域的 TiFlash 副本。如果无法访问所有的 TiFlash 数据，则查询将报错。
```

File: system-variables.md [hunk 5]
```diff
@@ -4172,6 +4172,21 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+- 特殊情况
+  - 如果 TiDB 节点未设置区域属性，并且 TiFlash 节点选择策略不是 "all_replicas"，则会忽略 TiFlash 节点选择策略，将使用所有 TiFlash 节点进行 TiFlash 查询。并且会有一个警告消息：The variable tiflash_replica_read is ignored。
```
Comment (joey-yez) [inline system-variables.md:4187]:
```suggestion
  - 如果 TiDB 节点未设置区域属性，并且 TiFlash 副本选择策略不是 "all_replicas"，则会忽略 TiFlash 副本选择策略，将使用所有 TiFlash 副本进行 TiFlash 查询。并且会有一个警告消息：The variable tiflash_replica_read is ignored。
```

File: system-variables.md [hunk 6]
```diff
@@ -4172,6 +4172,21 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+### tiflash_replica_read
+
+- 作用范围：SESSION | GLOBAL
+- 持久化至集群：是
+- 类型：枚举
+- 默认值："all_replicas"
+- 可选值："all_replicas"、"closest_adaptive"、"closest_replicas"
+- 该变量用于设置当查询需要使用 TiFlash 引擎时，TiFlash 节点的选择策略。
+  - "all_replicas" 表示使用所有可用节点进行分析计算。
+  - "closest_adaptive" 表示使用与入口 TiDB 相同区域的节点。如果无法访问所有的 TiFlash 数据，则查询将通过与入口 TiDB 相同区域的节点访问来自其他区域的 TiFlash 节点。
```
Comment (joey-yez) [inline system-variables.md:4184]:
```suggestion
  - "closest_adaptive" 表示尽量使用与入口 TiDB 相同区域的 TiFlash 副本进行分析计算。如果无法访问所有的 TiFlash 数据，则查询将使用与入口 TiDB 不同区域的 TiFlash 副本及对应的 TiFlash 节点。
```

File: system-variables.md [hunk 7]
```diff
@@ -4172,6 +4172,21 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
     * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
 - 理论上窗口函数的性能会随着该值的增加线性提升。但是如果设置的值超过实际的物理线程数，反而会导致性能下降。
```
Comment (qiancai) [inline system-variables.md:4174]: 关于 TiFlash 的副本选择策略是否需要在 https://docs.pingcap.com/zh/tidb/dev/use-tidb-to-read-tiflash)](https://docs.pingcap.com/zh/tidb/dev/use-tidb-to-read-tiflash 文档里也简单说明下呢，这样更方便用户知道这个功能

File: system-variables.md [hunk 8]
```diff
@@ -4172,6 +4172,21 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+- 默认值："all_replicas"
+- 可选值："all_replicas"、"closest_adaptive"、"closest_replicas"
+- 该变量用于设置当查询需要使用 TiFlash 引擎时，TiFlash 副本的选择策略。
+  - "all_replicas" 表示使用所有的 TiFlash 副本进行分析计算。
+  - "closest_adaptive" 表示尽量使用与入口 TiDB 相同区域的 TiFlash 副本进行分析计算。如果无法访问所有的 TiFlash 数据，则查询将使用与入口 TiDB 不同区域的 TiFlash 副本及对应的 TiFlash 节点。
+  - "closest_replicas" 表示仅使用与入口 TiDB 相同区域的 TiFlash 副本。如果无法访问所有的 TiFlash 数据，则查询将报错。
+- 特殊情况
+  - 如果 TiDB 节点未设置区域属性，并且 TiFlash 副本选择策略不是 "all_replicas"，则会忽略 TiFlash 副本选择策略，将使用所有 TiFlash 副本进行 TiFlash 查询。并且会有一个警告消息：The variable tiflash_replica_read is ignored。
+  - 如果 TiFlash 节点未设置区域属性，则将其视为不属于任何区域的节点。
+
```
Comment (qiancai) [inline system-variables.md:4189]: 顺便问下，这个区域属性是通过 https://docs.pingcap.com/zh/tidb/dev/schedule-replicas-by-topology-labels 设置的吗

File: system-variables.md (hunk 7 above)
Comment (joey-yez) [reply system-variables.md:4174]: 我建议在 https://docs.pingcap.com/zh/tidb/dev/create-tiflash-replicas#%E8%AE%BE%E7%BD%AE%E5%8F%AF%E7%94%A8%E5%8C%BA 这里增加一部分说明。上面的 link 文档主要是 tikv 和 tiflash 之间的选择，不太适合。

File: system-variables.md [hunk 9]
```diff
@@ -4172,6 +4172,21 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+- 类型：枚举型
+- 默认值：`all_replicas`
+- 可选值：`all_replicas`、`closest_adaptive`、`closest_replicas`
+- 该变量用于设置当查询需要使用 TiFlash 引擎时，TiFlash 副本的选择策略。
+  - `all_replicas` 表示使用所有的 TiFlash 副本进行分析计算。
+  - `closest_adaptive` 表示尽量使用与当前发起查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则再使用其他区域的 TiFlash 副本及对应的 TiFlash 节点。
+  - `closest_replicas` 表示仅使用与发起当前查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则查询将报错。
+- 特殊情况
+  - 如果 TiDB 节点未设置区域属性，并且 TiFlash 副本选择策略不是 `all_replicas` 时，TiFlash 引擎将忽略 TiFlash 副本选择策略，使用所有 TiFlash 副本进行 TiFlash 查询，并且返回警告 `The variable tiflash_replica_read is ignored`。
+  - 如果 TiFlash 节点未设置区域属性，则将其视为不属于任何区域的节点。
```
Comment (qiancai) [inline system-variables.md:4188]:
```suggestion

> **注意：**
> 
 > - 如果 TiDB 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tidb-的-labels可选)，并且 TiFlash 副本选择策略不是 `all_replicas` 时，TiFlash 引擎将忽略 TiFlash 副本选择策略，使用所有 TiFlash 副本进行 TiFlash 查询，并且返回警告 `The variable tiflash_replica_read is ignored`。
 > - 如果 TiFlash 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tikv-和-tiflash-的-labels)，则将其视为不属于任何区域的节点。
```

File: tiflash/create-tiflash-replicas.md [hunk 10]
```diff
@@ -250,3 +250,5 @@ SELECT TABLE_NAME FROM information_schema.tables where TABLE_SCHEMA = "<db_name>
     ```
 
 关于使用 label 进行副本调度划分可用区的更多内容，可以参考[通过拓扑 label 进行副本调度](/schedule-replicas-by-topology-labels.md)，[同城多数据中心部署 TiDB](/multi-data-centers-in-one-city-deployment.md) 与[两地三中心部署](/three-data-centers-in-two-cities-deployment.md)。
+
+TiFlash 提供不同可用区的副本访问策略，具体请参考变量 [tiflash_replica_read](/system-variables.md#tiflash_replica_read)。
```
Comment (qiancai) [inline tiflash/create-tiflash-replicas.md:254]: 前面提到的是区域，这里提到的是可用区，是否可以统一下用其中一个

File: system-variables.md [hunk 11]
```diff
@@ -4172,6 +4172,21 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
     * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
 - 理论上窗口函数的性能会随着该值的增加线性提升。但是如果设置的值超过实际的物理线程数，反而会导致性能下降。
 
+### `tiflash_replica_read` <span class="version-mark">从 v7.2.0 版本开始引入</span>
+
```
[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply system-variables.md:4176]: 是一个事情，就是 tiflash 副本放在哪里，用户的相关设置方式要做全套

[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply system-variables.md:4176]: 这里关注那个区域属性的文档就行

File: tiflash/create-tiflash-replicas.md (hunk 10 above)
Comment (XuHuaiyu) [inline tiflash/create-tiflash-replicas.md:254]:
```suggestion
TiFlash 提供不同区域的副本访问策略，具体请参考变量 [tiflash_replica_read](/system-variables.md#tiflash_replica_read)。
```

File: system-variables.md [hunk 12]
```diff
@@ -4172,6 +4172,23 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
     * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
 - 理论上窗口函数的性能会随着该值的增加线性提升。但是如果设置的值超过实际的物理线程数，反而会导致性能下降。
 
+### `tiflash_replica_read` <span class="version-mark">从 v7.2.0 版本开始引入</span>
```
Comment (qiancai) [inline system-variables.md:4175]:
```suggestion
### `tiflash_replica_read` <span class="version-mark">从 v7.3.0 版本开始引入</span>
```

File: system-variables.md [hunk 13]
```diff
@@ -4172,6 +4172,23 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+- 默认值：`all_replicas`
+- 可选值：`all_replicas`、`closest_adaptive`、`closest_replicas`
+- 该变量用于设置当查询需要使用 TiFlash 引擎时，TiFlash 副本的选择策略。
+  - `all_replicas` 表示使用所有的 TiFlash 副本进行分析计算。
+  - `closest_adaptive` 表示尽量使用与当前发起查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则再使用其他区域的 TiFlash 副本及对应的 TiFlash 节点。
+  - `closest_replicas` 表示仅使用与发起当前查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则查询将报错。
+
+> **注意：**
+> 
+ > - 如果 TiDB 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tidb-的-labels可选)，并且 TiFlash 副本选择策略不是 `all_replicas` 时，TiFlash 引擎将忽略 TiFlash 副本选择策略，使用所有 TiFlash 副本进行 TiFlash 查询，并且返回警告 `The variable tiflash_replica_read is ignored`。
```
Comment (ran-huang) [inline system-variables.md:4189]:
```suggestion
> - 如果 TiDB 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tidb-的-labels可选)，并且 TiFlash 副本选择策略不是 `all_replicas` 时，TiFlash 引擎将忽略 TiFlash 副本选择策略，使用所有 TiFlash 副本进行 TiFlash 查询，并且返回警告 `The variable tiflash_replica_read is ignored`。
```

File: system-variables.md [hunk 14]
```diff
@@ -4172,6 +4172,23 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+ > - 如果 TiFlash 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tikv-和-tiflash-的-labels)，则将其视为不属于任何区域的节点。
```
Comment (ran-huang) [inline system-variables.md:4190]:
```suggestion
> - 如果 TiFlash 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tikv-和-tiflash-的-labels)，则将其视为不属于任何区域的节点。
```

**Review by ran-huang — APPROVED** [review]

**Review by qiancai — APPROVED** [review]

File: tiflash/create-tiflash-replicas.md [hunk 15]
```diff
@@ -250,3 +250,5 @@ SELECT TABLE_NAME FROM information_schema.tables where TABLE_SCHEMA = "<db_name>
     ```
 
 关于使用 label 进行副本调度划分可用区的更多内容，可以参考[通过拓扑 label 进行副本调度](/schedule-replicas-by-topology-labels.md)，[同城多数据中心部署 TiDB](/multi-data-centers-in-one-city-deployment.md) 与[两地三中心部署](/three-data-centers-in-two-cities-deployment.md)。
+
+TiFlash 提供不同区域的副本访问策略，具体请参考变量 [tiflash_replica_read](/system-variables.md#tiflash_replica_read)。
```
Comment (qiancai) [inline tiflash/create-tiflash-replicas.md:254]:
```suggestion
TiFlash 提供不同区域的副本访问策略，具体请参考变量 [`tiflash_replica_read`](/system-variables.md#tiflash_replica_read-从-v730-版本开始引入)。
```

File: tiflash/create-tiflash-replicas.md [hunk 16]
```diff
@@ -250,3 +250,5 @@ SELECT TABLE_NAME FROM information_schema.tables where TABLE_SCHEMA = "<db_name>
     ```
 
 关于使用 label 进行副本调度划分可用区的更多内容，可以参考[通过拓扑 label 进行副本调度](/schedule-replicas-by-topology-labels.md)，[同城多数据中心部署 TiDB](/multi-data-centers-in-one-city-deployment.md) 与[两地三中心部署](/three-data-centers-in-two-cities-deployment.md)。
+
+TiFlash 提供不同区域的副本访问策略，具体请参考变量 [`tiflash_replica_read`](/system-variables.md#tiflash_replica_read-从-v730-版本开始引入)。
```
Comment (qiancai) [inline tiflash/create-tiflash-replicas.md:254]:
```suggestion
TiFlash 提供不同区域的副本选择策略，具体请参考变量 [`tiflash_replica_read`](/system-variables.md#tiflash_replica_read-从-v730-版本开始引入)。
```

File: tiflash/create-tiflash-replicas.md [hunk 17]
```diff
@@ -250,3 +250,5 @@ SELECT TABLE_NAME FROM information_schema.tables where TABLE_SCHEMA = "<db_name>
     ```
 
 关于使用 label 进行副本调度划分可用区的更多内容，可以参考[通过拓扑 label 进行副本调度](/schedule-replicas-by-topology-labels.md)，[同城多数据中心部署 TiDB](/multi-data-centers-in-one-city-deployment.md) 与[两地三中心部署](/three-data-centers-in-two-cities-deployment.md)。
+
+TiFlash 提供不同区域的副本选择策略，具体请参考变量 [`tiflash_replica_read`](/system-variables.md#tiflash_replica_read-从-v730-版本开始引入)。
```
Comment (qiancai) [inline tiflash/create-tiflash-replicas.md:254]:
```suggestion
TiFlash 支持设置不同区域的副本选择策略，具体请参考变量 [`tiflash_replica_read`](/system-variables.md#tiflash_replica_read-从-v730-版本开始引入)。
```

File: system-variables.md [hunk 18]
```diff
@@ -4172,6 +4172,23 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+
+- 作用范围：SESSION | GLOBAL
+- 持久化至集群：是
+- 类型：枚举型
+- 默认值：`all_replicas`
+- 可选值：`all_replicas`、`closest_adaptive`、`closest_replicas`
+- 该变量用于设置当查询需要使用 TiFlash 引擎时，TiFlash 副本的选择策略。
+  - `all_replicas` 表示使用所有的 TiFlash 副本进行分析计算。
+  - `closest_adaptive` 表示尽量使用与当前发起查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则再使用其他区域的 TiFlash 副本及对应的 TiFlash 节点。
+  - `closest_replicas` 表示仅使用与发起当前查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则查询将报错。
```
Comment (qiancai) [inline system-variables.md:4185]:
```suggestion
    - `all_replicas` 表示使用所有的 TiFlash 副本进行分析计算。
    - `closest_adaptive` 表示尽量使用与当前发起查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则再使用其他区域的 TiFlash 副本及对应的 TiFlash 节点。
    - `closest_replicas` 表示仅使用与发起当前查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则查询将报错。
```

**Review by joey-yez — APPROVED** [review]

**Pull request merged**